### ISSUE TITLE
`nuclei` fails to start when `~/.config` is not writable

### ISSUE THREAD

**Issue by ZhongRuoyu** [projectdiscovery/nuclei#3576]
### Nuclei version:

nuclei 2.9.2 (built from https://github.com/projectdiscovery/nuclei/archive/v2.9.2.tar.gz)

### Current Behavior:

`nuclei` fails to start when the `.config` directory under home directory is not writable:

```console
$ nuclei -target google.com -t test.yaml
[ERR] failed to create config directory at /Users/brew/.config/nuclei got: mkdir /Users/brew/.config/nuclei: operation not permitted
[ERR] failed to write config file at /Users/brew/.config/nuclei/.templates-config.json got: [:RUNTIME] could not create nuclei config directory at /Users/brew/.config/nuclei <- mkdir /Users/brew/.config/nuclei: operation not permitted

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.9.2

		projectdiscovery.io

[INF] nuclei-templates are not installed, installing...
[INF] nuclei-templates are not installed, installing...
[FTL] Could not create runner: [:RUNTIME] could not create config file <- open /Users/brew/.config/nuclei/reporting-config.yaml: no such file or directory
```

### Expected Behavior:

The `.config` directory under home directory can be unwritable for some reason. For instance, when `nuclei` is being run in a sandboxed environment, or when the `.config` is deliberately made read-only.

`nuclei` should be able to accept a config directory other than the default one; or alternatively, it should not always attempt at creating the directory at startup.

It would probably help to have a flag `-config-dir` or an environment variable `NUCLEI_CONFIG_DIR` that allows overriding the default config directory.

### Steps To Reproduce:

1. `chmod -rx ~/.config`
2. `nuclei -target google.com -t test.yaml`
3. See error.

### Anything else:

This error was observed while packaging `nuclei` for Homebrew in https://github.com/Homebrew/homebrew-core/pull/128819. Homebrew's tests are executed in a temporary sandboxed environment, where the `HOME` environment variable is changed to a temporary directory. However, [`getDefaultConfigDir`](https://github.com/projectdiscovery/nuclei/blob/v2.9.2/v2/pkg/catalog/config/nucleiconfig.go#L298)'s implementation [uses Go's os/user package](https://github.com/projectdiscovery/utils/blob/main/folder/folderutil.go#L143-L150) to determine the home directory; `os/user` retrieves the home directory information from the system interface ([`pwd.h`](https://manpages.ubuntu.com/manpages/jammy/man7/pwd.h.7posix.html)), which does not seem to respect the sandbox settings.

**Comment by ehsandeep** [comment]
@ZhongRuoyu, thanks for reporting this, there is already a flag for this specific case.

```
   -config-directory string       override the default config path ($home/.config)
````

**Comment by ehsandeep** [comment]
cc @tarunKoyalwar , `NUCLEI_CONFIG_DIR` could be a good addition for better UX

**Comment by ZhongRuoyu** [comment]
Thanks for your quick response @ehsandeep. May I know how it can be used? I am still able to reproduce the error with nuclei 2.9.2:

```console
$ ./nuclei start -target google.com -config-directory /tmp
[ERR] failed to create config directory at /home/ruoyu/.config/nuclei got: mkdir /home/ruoyu/.config/nuclei: permission denied
[ERR] failed to write config file at /home/ruoyu/.config/nuclei/.templates-config.json got: [:RUNTIME] could not create nuclei config directory at /home/ruoyu/.config/nuclei <- mkdir /home/ruoyu/.config/nuclei: permission denied

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.9.2

		projectdiscovery.io

[INF] Your current nuclei-templates  are outdated. Latest is v9.4.2
[FTL] Could not create runner: [:RUNTIME] could not create config file <- open /home/ruoyu/.config/nuclei/reporting-config.yaml: permission denied

$ ./nuclei -version
[ERR] failed to create config directory at /home/ruoyu/.config/nuclei got: mkdir /home/ruoyu/.config/nuclei: permission denied
[ERR] failed to write config file at /home/ruoyu/.config/nuclei/.templates-config.json got: [:RUNTIME] could not create nuclei config directory at /home/ruoyu/.config/nuclei <- mkdir /home/ruoyu/.config/nuclei: permission denied
[INF] Nuclei Engine Version: v2.9.2
```

If I understand it correctly, this `getDefaultConfigDir` invocation in the `init` function makes it so that `~/.config/nuclei` is always created unconditionally:

https://github.com/projectdiscovery/nuclei/blob/e3ce33afda548c67eaafe761bfeb9800b015e5d4/v2/pkg/catalog/config/nucleiconfig.go#L271-L277

https://github.com/projectdiscovery/nuclei/blob/e3ce33afda548c67eaafe761bfeb9800b015e5d4/v2/pkg/catalog/config/nucleiconfig.go#L298-L309

**Comment by tarunKoyalwar** [comment]
@ZhongRuoyu , value of userCfgDir should be
```go
userCfgDir , _ := os.UserConfigDIr()
```

**Comment by tarunKoyalwar** [comment]
@ZhongRuoyu yeah  we will sort it out with proper migration in next release . the change also need to be done upstream in `goflags` .

**Comment by Yablargo** [comment]
I'm going to open a PR just to start the dialog, but I've been working with nuclei in a lambda for a few months now and making fixes recommended in this thread worked great. I'd love to not have to use a custom fork of nuclei and just use the main branch.

**Comment by S4lt5** [comment]
I've gotta get my head around it, but the latest versions are back to not working with this change, I am starting to get template errors, so I really need to figure this out eventually. I think I am still stuck on 2.9.6 and am starting to get template errors

**Comment by tarunKoyalwar** [comment]
@S4lt5 since this is a breaking change we are hoping to release it with next major release . which will happen soon probably within a month

> but the latest versions are back to not working with this change
is doing this change https://github.com/projectdiscovery/nuclei/issues/3576#issuecomment-1516218619 on a nuclei work not resolving the issue ?? or are you perhaps facing other issue

**Comment by S4lt5** [comment]
The last time I checked, it failed in another area, but this is because my whole filesystem is read-only as I am running in a lambda function. Overall the experience is wonderful, i am able to scan 10s of thousands of sites in 10 minutes when running a single template, I just have a lot of backlog so I haven't really gotten around to fixing it yet.

**Comment by tarunKoyalwar** [comment]
@S4lt5 @ZhongRuoyu  with this [PR](https://github.com/projectdiscovery/nuclei/pull/4228) nuclei will use platform / sandbox specific config and cache directories and fixes this ~/.config not writeable issue

can you try nuclei from this PR and check if you are still facing any similar issue

**Comment by S4lt5** [comment]
Awesome, I will give it a shot as I am rebuilding in a new environment this week already.

**Comment by S4lt5** [comment]
This appears to have done the trick, doing a rather large scan and will see how my results turn up!

**Comment by S4lt5** [comment]
Seems great, thank you very much PD Team! I can drop my custom fork now, and rejoice!

**Comment by ZhongRuoyu** [comment]
Thanks @tarunKoyalwar and the Nuclei team! I gave the PR a try, and I can see that the issue is resolved. Closing now.
